arch/riscv/core/reset.S

```diff
@@ -58,36 +135,100 @@ boot_master_core:
 	 * all exception flags.
 	 */
 	fscsr x0, x0
-#endif
-
+#endif /* CONFIG_FPU */
 #ifdef CONFIG_INIT_STACKS
 	/* Pre-populate all bytes in z_interrupt_stacks with 0xAA */
 	la t0, z_interrupt_stacks
+#if CONFIG_SMP
+	li t1, CONFIG_ISR_STACK_SIZE * CONFIG_MP_TOTAL_NUM_CPUS
+	add t1, t1, t0
+	/* Only do this once on the startup hart */
+    li a1, CONFIG_HART_TO_USE
+    csrr a0, mhartid
+    bne a0, a1, skip_stack_fill
+#else /* CONFIG_SMP */
 	li t1, CONFIG_ISR_STACK_SIZE
 	add t1, t1, t0
+#endif /* CONFIG_SMP */
 
 	/* Populate z_interrupt_stacks with 0xaaaaaaaa */
 	li t2, 0xaaaaaaaa
 aa_loop:
 	sw t2, 0x00(t0)
 	addi t0, t0, 4
 	blt t0, t1, aa_loop
-#endif
+
+skip_stack_fill:
+#endif /* CONFIG_INIT_STACKS */
 
 	/*
 	 * Initially, setup stack pointer to
-	 * z_interrupt_stacks + CONFIG_ISR_STACK_SIZE
+	 * z_interrupt_stacks[cpu] + CONFIG_ISR_STACK_SIZE
+	 * 
+	 * We always allocate these stacks based on the real hart id as we need to
+	 * account for all harts when allocating an initial stack even if
+	 * they are not going to run Zephyr... 
 	 */
 	la sp, z_interrupt_stacks
 	li t0, CONFIG_ISR_STACK_SIZE
+#if defined(CONFIG_SMP)
+	csrr a0, mhartid
+	mul a1, a0, t0
+	add sp, sp, a1
+#endif /* CONFIG_SMP */
 	add sp, sp, t0
 
 	csrw mscratch, sp
 
 #ifdef CONFIG_WDOG_INIT
 	call _WdogInit
-#endif
+#endif /* CONFIG_WDOG_INIT */
+
+	li a1, CONFIG_HART_TO_USE
+	csrr a0, mhartid
+	beq a0, a1, boot_master_core
+
+#if CONFIG_SMP
+	/* Second stage of wake up - a2 already set up */
+	li a1, RV_WAKE_WAIT
+	sd a1, 0(a2)
+	li a0, RV_WAKE_GO
+#endif	/* CONFIG_SMP */
+	/*
+	 * Enable software interrupt
+	 */
+	li  t0, MIP_MSIP
+	csrrs x0, mie, t0
 
+loop_slave_core:
```

Comment:
i don't think we should use `master`/`slave` nomenclature in the new code. Maybe `main`/`secondary` core?